lib/globus/client/endpoint.rb

```diff
@@ -98,18 +83,69 @@ def paths
         end
       end
 
+      # @see #paths
+      def full_path
+        paths.last
+      end
+
       def path_segments
         [user_id, "work#{work_id}", "version#{work_version}"]
       end
 
       def objects
         # List files at an endpoint https://docs.globus.org/api/transfer/file_operations/#list_directory_contents
-        response = connection.get("#{transfer_path}/ls?path=#{paths.last}")
+        response = connection.get("#{transfer_path}/ls?path=#{full_path}")
         return JSON.parse(response.body) if response.success?
 
         UnexpectedResponse.call(response)
       end
 
+      def files
+        objects["DATA"].select { |object| object["DATA_TYPE"] == "file" }
+      end
+
+      def access_request(permissions:)
+        response = if access_rule_id
+          connection.put("#{access_path}/#{access_rule_id}") do |req|
+            req.body = {
+              DATA_TYPE: "access",
+              permissions:
+            }.to_json
+            req.headers["Content-Type"] = "application/json"
+          end
+        else
+          connection.post(access_path) do |req|
+            req.body = {
+              DATA_TYPE: "access",
+              principal_type: "identity",
+              principal: user,
+              path: full_path,
+              permissions:,
+              notify_email: "#{user_id}@stanford.edu"
+            }.to_json
+            req.headers["Content-Type"] = "application/json"
+          end
+        end
+
+        return response if response.success?
+
+        UnexpectedResponse.call(response)
+      end
```

Comment:
Now all access/permissions changes can handle updates as well as creates; we no longer assume that first we set to read-write and then to read-only.